Sources/Tools/plShaderAssembler/main.cpp

```diff
@@ -117,14 +117,36 @@ void ICreateHeader(const ST::string& varName, const plFileName& fileName, FILE*
     ST::printf(fp, "\t0x{x},", *codes++);
     ST::printf(fp, "\t0x{x},", *codes++);
     ST::printf(fp, "\t0x{x}\n", *codes++);
-    fputs("\t};", fp);
-    fputs("\n\n", fp);
+    ST::printf(fp, "\t];");
+    ST::printf(fp, "\n\n");
 
     ST::printf(fp, "static const plShaderDecl {}Decl(\"{}\", {}, {}byteLen, {}Codes);\n\n",
                varName, fileName, varName, varName, varName);
     ST::printf(fp, "static const plShaderRegister {}Register(&{}Decl);\n\n", varName, varName);
 }
 
+static bool AttemptToOpenOutFile(plFileName outFile, hsUNIXStream& outFp)
```

Comment:
```suggestion
static bool AttemptToOpenOutFile(const plFileName& outFile, hsUNIXStream& outFp)
```

No need to copy the file name. For bonus modernization, instead of returning a success value and taking a result reference, you could return an `std::optional<hsUNIXStream>`.

Reply:
attempted to have the function an std::optional like so

```
std::optional<hsUNIXStream> AttemptToOpenFile(const plFileName& file) {
    hsUNIXStream inFp;
    if (!inFp.Open(file, "rw")) {
        fputs("%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n", stderr);
        ST::printf(stderr, "Error opening file {}\n", file);
        fputs("%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\n", stderr);
        return std::nullopt; 
    }
    return inFp;
}
```
but the class hsUNIXStream doesnt seem to allow it.

Reply:
You may have to write a move constructor for `hsUNIXStream`.

Reply:
Other functions that return streams currently use `std::unique_ptr<hsStream>` or similar, which doesn't require move support.

Reply:
That's true, but here @jfmherokiller's snippet is moving an `hsUNIXStream` constructed on the stack into storage owned by the returned `std::optional<hsUNIXStream>`.

Reply:
using `std::optional<hsUNIXStream> inFp;.` seems very odd to me from a readability standpoint.